README.rst
Python Client for Google Cloud Scheduler API
============================================

|stable| |pypi| |versions|

`Google Cloud Scheduler API`_: lets you set up scheduled units of work to be executed at defined times or regular intervals. These work units are commonly known as cron jobs. Typical use cases might include sending out a report email on a daily basis, updating some cached data every 10 minutes, or updating some summary information once an hour.

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-scheduler.svg
   :target: https://pypi.org/project/google-cloud-scheduler/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-scheduler.svg
   :target: https://pypi.org/project/google-cloud-scheduler/
.. _Google Cloud Scheduler API: https://cloud.google.com/scheduler/docs
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/cloudscheduler/latest
.. _Product Documentation:  https://cloud.google.com/scheduler/docs

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Google Cloud Scheduler API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Google Cloud Scheduler API.:  https://cloud.google.com/scheduler/docs
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-scheduler


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-scheduler

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Google Cloud Scheduler API
   to see other available methods on the client.
-  Read the `Google Cloud Scheduler API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Google Cloud Scheduler API Product documentation:  https://cloud.google.com/scheduler/docs
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/google-cloud-scheduler/#history

## [2.11.2](https://github.com/googleapis/python-scheduler/compare/v2.11.1...v2.11.2) (2023-09-13)


### Documentation

* Minor formatting ([9e13935](https://github.com/googleapis/python-scheduler/commit/9e139351d958149032f3efa9120d9e8b8232b4e8))

## [2.11.1](https://github.com/googleapis/python-scheduler/compare/v2.11.0...v2.11.1) (2023-07-04)


### Bug Fixes

* Add async context manager return types ([#340](https://github.com/googleapis/python-scheduler/issues/340)) ([7e65978](https://github.com/googleapis/python-scheduler/commit/7e65978db72b38d1fed273562df86dd058cfd271))

## [2.11.0](https://github.com/googleapis/python-scheduler/compare/v2.10.0...v2.11.0) (2023-03-23)


### Features

* Location API methods ([#324](https://github.com/googleapis/python-scheduler/issues/324)) ([662e648](https://github.com/googleapis/python-scheduler/commit/662e6489710ea62b86b43d04f9bc69f9bc96e8b8))


### Documentation

* Fix formatting of request arg in docstring ([#328](https://github.com/googleapis/python-scheduler/issues/328)) ([94e406e](https://github.com/googleapis/python-scheduler/commit/94e406eb242b027824536e597ced29b6f82cba97))

## [2.10.0](https://github.com/googleapis/python-scheduler/compare/v2.9.1...v2.10.0) (2023-02-28)


### Features

* Enable "rest" transport in Python for services supporting numeric enums ([3de2adb](https://github.com/googleapis/python-scheduler/commit/3de2adbc753902bdfa72085567da4e45a520416e))

## [2.9.1](https://github.com/googleapis/python-scheduler/compare/v2.9.0...v2.9.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([1ebe016](https://github.com/googleapis/python-scheduler/commit/1ebe016e8c755cd94465090079a98d304ca0e730))


### Documentation

* Add documentation for enums ([1ebe016](https://github.com/googleapis/python-scheduler/commit/1ebe016e8c755cd94465090079a98d304ca0e730))

## [2.9.0](https://github.com/googleapis/python-scheduler/compare/v2.8.0...v2.9.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#307](https://github.com/googleapis/python-scheduler/issues/307)) ([455fa74](https://github.com/googleapis/python-scheduler/commit/455fa74db83f7ca93f822acec4002358cfd27f3e))

## [2.8.0](https://github.com/googleapis/python-scheduler/compare/v2.7.3...v2.8.0) (2022-12-14)


### Features

* Add support for `google.cloud.scheduler.__version__` ([ab1a908](https://github.com/googleapis/python-scheduler/commit/ab1a9089bbb8f9dd0b4ea26afffa8ae7e7ad069a))
* Add typing to proto.Message based class attributes ([ab1a908](https://github.com/googleapis/python-scheduler/commit/ab1a9089bbb8f9dd0b4ea26afffa8ae7e7ad069a))
* Updated Client Libraries for Cloud Scheduler ([#304](https://github.com/googleapis/python-scheduler/issues/304)) ([a6cad2f](https://github.com/googleapis/python-scheduler/commit/a6cad2f03d2846f672f8403d38d2fcb9da69912b))


### Bug Fixes

* Add dict typing for client_options ([ab1a908](https://github.com/googleapis/python-scheduler/commit/ab1a9089bbb8f9dd0b4ea26afffa8ae7e7ad069a))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([75a677a](https://github.com/googleapis/python-scheduler/commit/75a677a30af56568ae365715056360d23f4b7f7c))
* Drop usage of pkg_resources ([75a677a](https://github.com/googleapis/python-scheduler/commit/75a677a30af56568ae365715056360d23f4b7f7c))
* Fix timeout default values ([75a677a](https://github.com/googleapis/python-scheduler/commit/75a677a30af56568ae365715056360d23f4b7f7c))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([ab1a908](https://github.com/googleapis/python-scheduler/commit/ab1a9089bbb8f9dd0b4ea26afffa8ae7e7ad069a))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([75a677a](https://github.com/googleapis/python-scheduler/commit/75a677a30af56568ae365715056360d23f4b7f7c))

## [2.7.3](https://github.com/googleapis/python-scheduler/compare/v2.7.2...v2.7.3) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#291](https://github.com/googleapis/python-scheduler/issues/291)) ([f0195d0](https://github.com/googleapis/python-scheduler/commit/f0195d0f5ef9e8b6342da965832a2a93fe795df2))

## [2.7.2](https://github.com/googleapis/python-scheduler/compare/v2.7.1...v2.7.2) (2022-10-03)


### Bug Fixes

* **deps:** Require protobuf >= 3.20.2 ([#288](https://github.com/googleapis/python-scheduler/issues/288)) ([c8357fe](https://github.com/googleapis/python-scheduler/commit/c8357fe07bd79d52b72f2733b7a7cf9557386b57))

## [2.7.1](https://github.com/googleapis/python-scheduler/compare/v2.7.0...v2.7.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#268](https://github.com/googleapis/python-scheduler/issues/268)) ([7081d77](https://github.com/googleapis/python-scheduler/commit/7081d777761e201a29b3d6d0542e30a1761350e0))
* **deps:** require proto-plus >= 1.22.0 ([7081d77](https://github.com/googleapis/python-scheduler/commit/7081d777761e201a29b3d6d0542e30a1761350e0))

## [2.7.0](https://github.com/googleapis/python-scheduler/compare/v2.6.4...v2.7.0) (2022-07-16)


### Features

* add audience parameter ([c8adf9c](https://github.com/googleapis/python-scheduler/commit/c8adf9c9877d4bbea2f5b282f95cb9f56011a94f))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#258](https://github.com/googleapis/python-scheduler/issues/258)) ([a57f965](https://github.com/googleapis/python-scheduler/commit/a57f96505640623170ae6b86c604127d14481561))
* require python 3.7+ ([#256](https://github.com/googleapis/python-scheduler/issues/256)) ([6b0faa0](https://github.com/googleapis/python-scheduler/commit/6b0faa00c155798fac1218a1f05cda54b3651f65))

## [2.6.4](https://github.com/googleapis/python-scheduler/compare/v2.6.3...v2.6.4) (2022-06-06)


### Bug Fixes


owl-bot-staging/v1beta1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/scheduler_v1beta1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
